lib/portlayer/exec2/container.go

```diff
@@ -0,0 +1,102 @@
+// Copyright 2016 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package exec2
+
+import (
+	"time"
+
+	"github.com/vmware/vic/pkg/vsphere/vm"
+)
+
+// All data in this struct must be data that is either immutable
+// or can be relied upon without having to query either the container guest
+// or the underlying infrastructure. Some of this state will be updated by events
+type container struct {
+	ConstantConfig
+
+	vm       vm.VirtualMachine
+	runState RunState
+
+	config          Config
+	processCfg      ProcessConfig // container main process
+	childProcessCfg []ProcessConfig
+
+	filesToCopy []FileToCopy // cache if copy while stopped
+}
+
+// config that will be applied to a container on commit
+// Needs to be public as it will be shared by net, storage etc
+type PendingCommit struct {
+	ConstantConfig
+
+	runState       RunState
+	config         Config
+	processConfig  ProcessConfig
+	childProcesses []ProcessConfig
+	filesToCopy    []FileToCopy
+}
+
+// config state that cannot change for the lifetime of the container
+type ConstantConfig struct {
+	CID     ID
+	Created time.Time
+	Limits  ResourceLimits
+}
+
+// variable container configuration state
+type Config struct {
+	Name    string
+	WorkDir string
+}
+
+// configuration state of a container process
+type ProcessConfig struct {
+	PID      ID
+	ExecPath string
+	ExecArgs string
+}
+
+func NewProcessConfig(execPath string, execArgs string) ProcessConfig {
+	return ProcessConfig{PID: GenerateID(), ExecArgs: execArgs, ExecPath: execPath}
+}
+
+type ProcessStatus int
+
+const (
+	Started = iota
+	Exited
+)
+
+// runtime status of a container process
+type ProcessRunState struct {
```

Comment:
Does only process need a run state?  What about a container?  How do we represent exit error and message?  Just logs?

In the world of Rkt and Kubernetes, there is some orchestration within containers or pods.  In docker's world view (maybe it has changed?), orchestration is done outside of containers.  Maybe we should allow for both world view.  Allowing for orchestration within containers would be more efficient.  It's a difficult question to answer.  It really depends on the direction of containers in the industry.  One container per process certainly has their pros and cons.  The question is where the industry wants to go.  Or maybe this is one of the future defining feature of VIC?